concordium-node/src/grpc2.rs

```diff
@@ -2615,3 +2646,124 @@ where
+    fn poll_read(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        std::pin::Pin::new(&mut self.addr).poll_read(cx, buf)
+    }
+}
+
+/// Forward implementation to that of the inner [`AddrStream`].
+impl tokio::io::AsyncWrite for AddrStreamWithTicket {
+    fn poll_write(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, std::io::Error>> {
+        std::pin::Pin::new(&mut self.addr).poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        std::pin::Pin::new(&mut self.addr).poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        std::pin::Pin::new(&mut self.addr).poll_shutdown(cx)
+    }
+}
+
+/// Forward implementation to that of the inner [`AddrStream`].
+impl Connected for AddrStreamWithTicket {
+    type ConnectInfo = <AddrStream as Connected>::ConnectInfo;
+
+    fn connect_info(&self) -> Self::ConnectInfo { self.addr.connect_info() }
+}
+
+impl futures::Stream for ConnStreamWithTicket {
+    type Item = Result<AddrStreamWithTicket, std::io::Error>;
+
+    // This is where we limit the number of connections.
+    // We only return a new item if we can combine it with a permit.
+    //
+    // The choice made here is that we don't even try to accept new connections
+    // unless we have a permit. The connections will thus remain in the accept queue
+    // maintained by the OS. An alternative choice would have been to accept a
+    // connection, then check if we have a permit, and then drop the connection
+    // if we don't. This would be better in some circumstances, but it would
+    // also mean that if there is a short burst of short-lived connections they
+    // would not be handled, whereas with the current approach they do.
+    //
+    // The best option would be to have a full backlog handling with expiry, and
+    // some rate limiting based on IP (i.e., no more than X connections per IP).
+    // However that is a more extensive change. The primary purpose of the current
+    // solution is to ensure the node always has enough resources. Not load
+    // balancing.
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        // Wait until a slot is available.
+        let Some(permit) = futures::ready!(self.semaphore.poll_acquire(cx)) else {
+            log::error!("Semaphore unexpectedly dropped. Stopping receiving new connections.");
+            return std::task::Poll::Ready(None);
```

Comment:
How is this handled by the 'server'? will it shutdown the grpc gracefully here or panic?

Reply:
It will not panic. It will do exactly what the `error` message says, it will not receive any more new connections.

Reply:
I added a comment about this.